### PR TITLE
Add ability to speak character when cursor routing

### LINKED ISSUE #8072: Braille issue: the letter in a text is not being reported by speech when moving the cursor to it with the cursor routing of the braille display
### Steps to reproduce:
1. Open a text editor
2. Write a sentence
3. Press the cursor routing of a braille display to move the cursor to a letter.

### Expected behavior:
NVDA should report that letter by speech, if speech is enabled.

### Actual behavior:
NVDA does not report the letter.

### System configuration:
NVDA version:
NVDA 2018.1

NVDA Installed or portable:
Both

Other information:

Windows version:
Windows 10.16299.248

Name and version of other software in use when reproducing the issue:
Text editor, MS Word, MS Outlook.

### Other questions:

Does the issue still occur after restarting your PC?
yes

Have you tried any other versions of NVDA?
2017.4

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #8072 

### Summary of the issue:
When cursor routing, it can sometimes be handy if speech follows along and tells you what character you moved to, especially when the character in braille is unknown to you.

### Description of user facing changes
* Added a new braille option "Speak character when routing cursor in text", disabled by default. When enabled, NVDA will speak the character you routed to after a cursor routing action.

### Description of development approach
Added a new option to the config spec, as well as a function to braille to speak the character at the given position. Also added an item to the settings panel as well as an unbound global command.

### Testing strategy:
- [x] Tested that NVDA will speak the character ath the cursor when routing when the option is on
- [x] Tested that NVDA won't speak the character ath the cursor when routing when the option is off
- [x] Tested that the global command (when bound) appropriately toggles the option

### Known issues with pull request:
None known

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
	- Introduced a setting to automatically speak the character at the cursor during braille cursor routing, enhancing auditory feedback for improved accessibility.
	- Added a checkbox in settings for users to enable or disable the new speech feature.
	- Implemented a toggle command for the speech feature in braille mode.

- **Documentation**
	- Updated user guide to include instructions for the new braille speech feature and customization options. 
	- Added notes in the change log regarding default settings and operational benefits for users. 

- **Bug Fixes**
	- Enhanced configuration management for better user interaction with the new settings.

### CODE REVIEW

File: source/globalCommands.py [hunk 1]
```diff
@@ -3593,6 +3593,22 @@ def script_braille_toggleShowCursor(self, gesture):
...
+	@gui.blockAction.when(gui.blockAction.Context.BRAILLE_MODE_SPEECH_OUTPUT)
+	def script_braille_toggleSpeakOnRouting(self, gesture):
+		state = config.conf["braille"]["speakOnRouting"] = not config.conf["braille"]["speakOnRouting"]
+		if state:
+			# Translators: The message announced when toggling on speaking character when routing.
+			state = _("Speak character when routing cursor in text on")
+		else:
+			# Translators: The message announced when toggling off speaking character when routing.
+			state = _("Speak character when routing cursor in text off")
+		ui.message(state)
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/globalCommands.py:3610]: @coderabbitai can you investgiate whether the pattern you suggest is used anywhere else in globalCommands? I think it is unnecessary to do this.

File: source/gui/settingsDialogs.py [hunk 2]
```diff
@@ -4447,6 +4447,14 @@ def makeSettings(self, settingsSizer):
 		)
 		self.bindHelpEvent("BrailleFormattingDisplay", self.formattingDisplayCombo)
 
+		# Translators: The label for a setting in braille settings to speak the character under the cursor when cursor routing in text.
+		speakOnRoutingText = _("Spea&k character when routing cursor in text")
+		self.speakOnRoutingCheckBox = followCursorGroupHelper.addItem(
+			wx.CheckBox(self, label=speakOnRoutingText),
+		)
+		self.bindHelpEvent("BrailleSettingsSpeakOnRouting", self.speakOnRoutingCheckBox)
+		self.speakOnRoutingCheckBox.Value = config.conf["braille"]["speakOnRouting"]
```
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/gui/settingsDialogs.py:4456]: @coderabbitai what would unit tests look like?

[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/gui/settingsDialogs.py:4456]: @coderabbitai this is unnecessary.

File: source/globalCommands.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (LeonarddeR) [reply source/globalCommands.py:3610]: @ocderabbitai Ok, I won't follow this up then.

**Comment by XLTechie** [comment]
@LeonarddeR

>  *  Added a new braille option "Speak character when routing cursor in text", disabled by default. When enabled, NVDA will speak the character you routed to

Given how many options for controlling small aspects there are (ever growing), and how few users go through after every update and look at new ones, or read through the changes:

wouldn't it make more sense to turn this on by default, let the new feature be exposed, and let those who don't like it figure out how to turn it off, instead of the other way around?

I too often run into the situation where I hear users say "I didn't know NVDA could even do that", just because we are too conservative and leave new features disabled by default. We assume that users will happen across them, recognize what they actually do, and recognize that what is done is something that they wanted all along.

I for one, don't consider those safe assumptions.

**Comment by LeonarddeR** [comment]
@XLTechie
 wrote:
> Given how many options for controlling small aspects there are (ever growing), 

This is certainly something to consider.

> wouldn't it make more sense to turn this on by default, let the new feature be exposed, and let those who don't like it figure out how to turn it off, instead of the other way around? 

No feedback on cursor routing has been the default since the start of NVDA and none has ever bothered to implement this feature until now. Personally, I would only enable the feature if, for example, I was working with foreign language braille tables. So to keep it short, I'm really reluctant to enable this by default.
What do you think @Adriani90?

**Comment by Qchristensen** [comment]
The how we should get new features in front of users is a very valuable discussion - although out of scope for this PR - so I created issue #16961 for it - @XLTechie and @LeonarddeR I copied your comments over and tagged you both in that issue.

**Comment by LeonarddeR** [comment]
@burmancomp could you please open a new issue for this? I think it could be an useful feature indeed, though much harder to implement.

File: source/globalCommands.py [hunk 3]
```diff
@@ -3593,6 +3593,22 @@ def script_braille_toggleShowCursor(self, gesture):
...
+	)
+	@gui.blockAction.when(gui.blockAction.Context.BRAILLE_MODE_SPEECH_OUTPUT)
+	def script_braille_toggleSpeakOnRouting(self, gesture):
+		state = config.conf["braille"]["speakOnRouting"] = not config.conf["braille"]["speakOnRouting"]
+		if state:
+			# Translators: The message announced when toggling on speaking character when routing.
+			state = _("Speak character when routing cursor in text on")
+		else:
+			# Translators: The message announced when toggling off speaking character when routing.
+			state = _("Speak character when routing cursor in text off")
```
Comment (seanbudd) [inline source/globalCommands.py:3609]: would it be nicer UX to put the state earlier?

```suggestion
			# Translators: The message announced when toggling on speaking character when routing.
			state = _("Toggled on speak character when routing cursor in text")
		else:
			# Translators: The message announced when toggling off speaking character when routing.
			state = _("Toggled off speak character when routing cursor in text")
```

File: source/gui/settingsDialogs.py [hunk 4]
```diff
@@ -4447,6 +4447,14 @@ def makeSettings(self, settingsSizer):
 		)
 		self.bindHelpEvent("BrailleFormattingDisplay", self.formattingDisplayCombo)
 
+		# Translators: The label for a setting in braille settings to speak the character under the cursor when cursor routing in text.
+		speakOnRoutingText = _("Spea&k character when routing cursor in text")
+		self.speakOnRoutingCheckBox = followCursorGroupHelper.addItem(
+			wx.CheckBox(self, label=speakOnRoutingText),
+		)
+		self.bindHelpEvent("BrailleSettingsSpeakOnRouting", self.speakOnRoutingCheckBox)
```
Comment (seanbudd) [inline source/gui/settingsDialogs.py:4455]:
```suggestion
		self.bindHelpEvent("BrailleSpeakOnRouting", self.speakOnRoutingCheckBox)
```

File: user_docs/en/userGuide.md [hunk 5]
```diff
@@ -2250,6 +2250,12 @@ The following symbols are defined:
 | ⠎ ("s")| Start strikethrough |
 | ⡎ ("s" with dot 7) | End strikethrough |
 
+##### Speak character when routing cursor in text {#BrailleSettingsSpeakOnRouting}
```
Comment (seanbudd) [inline user_docs/en/userGuide.md:2253]:
```suggestion
##### Speak character when routing cursor in text {#BrailleSpeakOnRouting}
```

**Review by seanbudd — APPROVED** [review]

File: source/globalCommands.py [hunk 6]
```diff
@@ -3593,6 +3593,22 @@ def script_braille_toggleShowCursor(self, gesture):
...
+		# Translators: Input help mode message for speak on routing command.
+		description=_("Toggles speaking the character under the cursor when routing cursor in text"),
+		category=SCRCAT_BRAILLE,
+	)
+	@gui.blockAction.when(gui.blockAction.Context.BRAILLE_MODE_SPEECH_OUTPUT)
+	def script_braille_toggleSpeakOnRouting(self, gesture):
+		state = config.conf["braille"]["speakOnRouting"] = not config.conf["braille"]["speakOnRouting"]
+		if state:
+			# Translators: The message announced when toggling on speaking character when routing.
+			state = _("Toggled on speak character when routing cursor in text")
```
Comment (XLTechie) [inline source/globalCommands.py:3606]: Wouldn't "Turned" be a less awkward word in English? Not sure how well it translates...

```suggestion
			state = _("Turned on speak character when routing cursor in text")
```

File: source/globalCommands.py [hunk 7]
```diff
@@ -3593,6 +3593,22 @@ def script_braille_toggleShowCursor(self, gesture):
...
+		else:
+			# Translators: The message announced when toggling off speaking character when routing.
+			state = _("Toggled off speak character when routing cursor in text")
```
Comment (XLTechie) [inline source/globalCommands.py:3609]:
```suggestion
			state = _("Turned off speak character when routing cursor in text")
```

File: source/globalCommands.py (hunk 6 above)
Comment (LeonarddeR) [reply source/globalCommands.py:3606]: I don't think we use the word `turn` that way somewhere else.
```suggestion
			state = _("Enabled speak character when routing cursor in text")
```

**Comment by LeonarddeR** [comment]
@Qchristensen Turns out there's some debate about wording. Could you chime in? Basically:
1. Toggle on/off
2. Turn on/off
3. Enable/disable

**Comment by seanbudd** [comment]
I think enabled/disabled is generally what we use, I just couldn't think of it at the time

**Comment by XLTechie** [comment]
@LeonarddeR I like The "Enabled/Disabled" idea better than either toggled or turned.

I retract my prior suggestion.

File: user_docs/en/changes.md [hunk 8]
```diff
@@ -15,6 +15,8 @@ The available options are:
   * Tags: Uses start and end tags to denote where certain font attributes begin and end. (#16864)
 * When the "Read by paragraph" option is enabled, NVDA can now be configured to indicate the start of paragraphs in braille. (#16895,  @nvdaes)
 * The timeout to perform a multiple keypress is now configurable; this may be especially useful for people with dexterity impairment. (#11929, @CyrilleB79)
+* When performing a braille cursor routing action, NVDA can now automatically speak the character ath the cursor. (#8072, @LeonarddeR)
```
Comment (Qchristensen) [inline user_docs/en/changes.md:18]:
```suggestion
* When performing a braille cursor routing action, NVDA can now automatically speak the character at the cursor. (#8072, @LeonarddeR)
```

(Removed extra 'h' in the word "at")

**Review by Qchristensen — CHANGES_REQUESTED** [review]
User Guide reads well, just made one small fix in the changes text

**Pull request merged**